module/zfs/zfs_vnops.c

```diff
@@ -979,13 +981,497 @@ zfs_get_done(zgd_t *zgd, int error)
+	    &dstzp->z_pflags, 8);
+
+	zilog = dstzfsvfs->z_log;
+
+	uid = KUID_TO_SUID(ZTOUID(dstzp));
+	gid = KGID_TO_SGID(ZTOGID(dstzp));
+	projid = dstzp->z_projid;
+
+	/*
+	 * Clone the file in reasonable size chunks.  Each chunk is cloned
+	 * in a separate transaction; this keeps the intent log records small
+	 * and allows us to do more fine-grained space accounting.
+	 */
+	while (length > 0) {
+		size = MIN(srcblksz * BRT_NBLOCKS, length);
+
+		if (zfs_id_overblockquota(dstzfsvfs, DMU_USERUSED_OBJECT,
+		    uid) ||
+		    zfs_id_overblockquota(dstzfsvfs, DMU_GROUPUSED_OBJECT,
+		    gid) ||
+		    (projid != ZFS_DEFAULT_PROJID &&
+		    zfs_id_overblockquota(dstzfsvfs, DMU_PROJECTUSED_OBJECT,
+		    projid))) {
+			error = SET_ERROR(EDQUOT);
+			break;
+		}
+
+		/*
+		 * Start a transaction.
+		 */
+		tx = dmu_tx_create(dstzfsvfs->z_os);
+
+		error = dmu_brt_readbps(srczfsvfs->z_os, srczp->z_id, srcoffset,
+		    size, tx, &bps, &nbps);
+		if (error != 0) {
+			dmu_tx_abort(tx);
+			break;
+		}
+		/*
+		 * Encrypted data is fine as long as it comes from the same
+		 * dataset.
+		 * TODO: We want to extend it in the future to allow cloning to
+		 * datasets with the same keys, like clones or to be able to
+		 * clone a file from a snapshot of an encrypted dataset into the
+		 * dataset itself.
+		 */
+		if (BP_IS_PROTECTED(&bps[0])) {
+			if (srczfsvfs != dstzfsvfs) {
+				dmu_tx_abort(tx);
+				kmem_free(bps, sizeof (bps[0]) * nbps);
+				error = SET_ERROR(ENODEV);
+				break;
+			}
+		}
+
+		dmu_tx_hold_sa(tx, dstzp->z_sa_hdl, B_FALSE);
+		db = (dmu_buf_impl_t *)sa_get_db(dstzp->z_sa_hdl);
+		DB_DNODE_ENTER(db);
+		dmu_tx_hold_write_by_dnode(tx, DB_DNODE(db), dstoffset, size);
+		DB_DNODE_EXIT(db);
+		zfs_sa_upgrade_txholds(tx, dstzp);
+		error = dmu_tx_assign(tx, TXG_WAIT);
+		if (error != 0) {
+			dmu_tx_abort(tx);
+			kmem_free(bps, sizeof (bps[0]) * nbps);
+			break;
+		}
+
+		/*
+		 * Copy source znode's block size. This only happens on the
+		 * first iteration since zfs_rangelock_reduce() will shrink down
+		 * lr_length to the appropriate size.
+		 */
+		if (dstlr->lr_length == UINT64_MAX) {
+			zfs_grow_blocksize(dstzp, srcblksz, tx);
+			zfs_rangelock_reduce(dstlr, dstoffset, length);
+		}
+
+		dmu_brt_addref(dstzfsvfs->z_os, dstzp->z_id, dstoffset, size,
+		    tx, bps, nbps);
+
+		zfs_clear_setid_bits_if_necessary(dstzfsvfs, dstzp, cr,
+		    &clear_setid_bits_txg, tx);
+
+		zfs_tstamp_update_setup(dstzp, CONTENT_MODIFIED, mtime, ctime);
+
+		/*
+		 * Update the file size (zp_size) if it has changed;
+		 * account for possible concurrent updates.
+		 */
+		while ((dstsize = dstzp->z_size) < dstoffset + size) {
+			(void) atomic_cas_64(&dstzp->z_size, dstsize,
+			    dstoffset + size);
+		}
+
+		error = sa_bulk_update(dstzp->z_sa_hdl, bulk, count, tx);
+
+		zfs_log_clone(zilog, tx, TX_CLONE, dstzp, dstioflag, dstoffset,
+		    size, srcblksz, bps, nbps);
+		dmu_tx_commit(tx);
+
+		kmem_free(bps, sizeof (bps[0]) * nbps);
```

Comment:
Why don't we provide the bps buffer to dmu_read_bps? That way, we don't alloc+free in every loop iteration

Reply:
Hmm, good idea. Done.